corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        self.logger.add_changes({'language': language})
+
+    def update_email(self, email):
+        self.user.email = email.lower()
+        self.logger.add_changes({'email': self.user.email})
+
+    def update_status(self, is_active):
+        self.user.is_active = is_active
+        self.logger.add_changes({'is_active': is_active})
+
+    def update_locations(self, location_codes, domain_info):
+        from corehq.apps.user_importer.importer import (
+            check_modified_user_loc,
+            find_location_id,
+            get_location_from_site_code
+        )
+
+        location_ids = find_location_id(location_codes, domain_info.location_cache)
+        users_current_primary_location_id = self.user.location_id
+        locations_updated, primary_loc_removed = check_modified_user_loc(location_ids,
+                                                                         self.user.location_id,
+                                                                         self.user.assigned_location_ids)
+        if primary_loc_removed:
+            self.user.unset_location(commit=False)
+        if locations_updated:
+            self.user.reset_locations(location_ids, commit=False)
+            self.logger.add_changes({'assigned_location_ids': location_ids})
+            if location_ids:
+                location_names = [get_location_from_site_code(code, domain_info.location_cache).name
+                                  for code in location_codes]
+                self.logger.add_info(_(f"Assigned locations: {location_names}"))
+        # log this after assigned locations are updated, which can re-set primary location
+        if self.user.location_id != users_current_primary_location_id:
+            self.logger.add_changes({'location_id': self.user.location_id})
+            if self.user.location_id:
+                users_updated_primary_location_name = get_user_primary_location_name(self.user, self.user_domain)
+                self.logger.add_info(
+                    _(f"Primary location: {users_updated_primary_location_name}"))
+
+    def update_role(self, role, domain_info):
+        role_qualified_id = domain_info.roles_by_name[role]
+        user_current_role = self.user.get_role(domain=self.user_domain)
+        self.role_updated = not (user_current_role
+                                 and user_current_role.get_qualified_id() == role_qualified_id)
+        if self.role_updated:
+            self.user.set_role(self.user_domain, role_qualified_id)
+
+    def save(self):
+        # Tracking for role is done post save to have role setup correctly on save
+        if self.role_updated:
+            new_role = self.user.get_role(domain=self.user_domain)
+            if new_role:
+                self.logger.add_info(_(f"Role: {new_role.name}[{new_role.get_id}]"))
+            else:
+                self.logger.add_change_message("Role: None")
+
+        # ToDo: save log before saving user
```

Comment:
is this todo still valid?

Reply:
I kept a note for me to reconsider this. If the user is saved, but logs are not, we would not have the history. Though the reverse is harmful as well.
Any thoughts about that?
Is this okay as it is, or better to put this in a transaction?

Reply:
Couch doesn't have transactions. I think the best option is to save the user first and then the log. If the log fails it should report to Sentry.

Reply:
Cool,

removed ToDo https://github.com/dimagi/commcare-hq/pull/29935/commits/091fa00a9422f1837f4a9e535f1f2d5c2ca25d0c